### ISSUE TITLE
[all] Support multiple distribution formats

### ISSUE THREAD

**Issue by postspectacular** [thi-ng/umbrella#23]
As per [Discord chat](https://discordapp.com/channels/445761008837984256/445761258696736790?jump=449196939200954398), we should look into easy-to-maintain solutions supporting distributions in different flavors.

Proposed initial solution to follow along these lines:
https://www.reddit.com/r/javascript/comments/6ngjkk/how_to_distribute_library_written_with_es6/dka3g5u/

**Comment by allforabit** [comment]
Hi @postspectacular I've been taking a look at this today. I'm just wondering for the first pass, will I go for something like the redux approach? You can see the layout of the package that is delivered to apps here: https://unpkg.com/redux@4.0.0/ The one exception might be to leave the es6 module version untranspiled to es5. 

It looks they include a fully self contained bundled version as well (both minified and unminified). I think this helps for tools like unpkg.com and for people who just want to include the file directly in the html page. It would also make it easier to use it with observablehq.com. You can test it here https://beta.observablehq.com/@tmcw/module-require-debugger

The bundling tool of choice for libs seems to be rollup: https://medium.com/webpack/webpack-and-rollup-the-same-but-different-a41ad427058c so I could probably use this with the rollup typescript plugin: https://github.com/rollup/rollup-plugin-typescript

**Comment by postspectacular** [comment]
Thanks for these new links! This constantly changing tool-of-the-month behavior of the JS community makes me feel like a complete beginner again. I've never used rollup and therefore can't comment on its merits, but will check it out ASAP and in the meantime I'd defer to your expertise to try it out and see if you can make it work. Maybe try it with a single module first (e.g. @thi.ng/checks - has no deps) on a new feature branch. Then I can clone it and do a test publish with version qualifier and we can test it in situ...

Ps. It seems neither unpkg.com nor that module-require-debugger seem to like scoped packages 😭

**Comment by postspectacular** [comment]
Oki, ignore my last comment about scoped packages. It works now, but didn't earlier. Likely there's a time delay for importing still unknown packages into unpkg...

**Comment by postspectacular** [comment]
Oki, I just tried out rollup with one of modules, but I already can see an a major issue with import paths:

I've configured @thi.ng/checks to use 2 separate tsconfig files, each with their own targets and building this works completely fine.

- CommonJS => `./lib`
- ES6 => "./es6"

```js
// package.json
{
...
"main": "./lib/index.js"
"module": "./es6/index.js"
...
}
```

However, I've no idea how to change all the imports in other places, neither of these works:

```ts
import { isArray } from "@thi.ng/checks/is-array";
import { isArray } from "@thi.ng/checks/lib/is-array";
import { isArray } from "@thi.ng/checks/es6/is-array";
```

**Comment by allforabit** [comment]
Would leaving the es6 files in the root of the folder help I wonder? It might be a little confusing as far as a folder layout goes but if it makes it work maybe this would be the best approach. I'm also wondering if it's necessary to use rollup for these two versions. Could it not just be compiled with typescript similarly to how it's already being done except with two tsconfigs for es6 and es5? Then rollup could be used to create a fully bundled version that could be put in the "dist" folder.

**Comment by postspectacular** [comment]
@allforabit I've tried it and I think the issue is has to do that I've switched the Typescript output module format from "commonjs" to "es2015". this alone seems to make TS not find the module anymore when referred to from other projects. I still have to try the combination where I keep the CommonJS output in the project root (as has been the case for now)...

As for rollup, yep, this wasn't part of the experiment so far - the way I understand it we would only need this for (additionally) providing UMD versions, correct?

**Comment by postspectacular** [comment]
We might need to explicitly set the [TS module resolution strategy](https://www.typescriptlang.org/docs/handbook/module-resolution.html#module-resolution-strategies says) to "node" when using "es2015" module format. Will try that tonight...

**Comment by allforabit** [comment]
Ah I see that makes sense. Yes rollup would just be used for the UMD version. Changing to the "node" resolution strategy seems like the solution to the issue, well spotted! It seems the classic resolution strategy is just their for backwards compatibility. 

Just to note, if you're trying to compile the more complex modules such as rstream there are some settings that I changed to get it working. Here's what worked for me: `tsc --outdir lib --target ES5 --lib ES6,DOM --downlevelIteration`. I'm not sure if polyfills would need to be added by the end users for things like Symbols. I'll check it with my react native project. It's very straightforward though to include these shims for the end user so it's probably not something that the library should include.

**Comment by postspectacular** [comment]
Hmm... I'd really prefer to avoid ES5 as default output since a lot of these polyfills will have negative impact on performance. Can people not just transpile this themselves if they really still need ES5?

**Comment by allforabit** [comment]
It would probably bloat the packages as well so yes it definitely shouldn't be the default output. So there wouldn't be an es5 version published at all? I guess that would be the simplest option. I think I misunderstood that this issue was about making an es5 version available. I'm not sure if just changing the module style from cjs to es will help with the issues outlined on the Discord channel:

> Thanks for the reply! I could transpile the code however it gets complicated. For example the Atom pkg lives under node_modules/@thi.ng/Atom. In my project, the tsoconfig sets target: est5 (to work with ie11) but I use exclude: "node_moduels" since I don't want to run my entire node_modules folder through the TS compiler. I make an exception in my include: "include": ["src/**/*", "node_modules/@thi.ng/**/*"]. But now in my outDir,  I have a node_modules folder. Now I believe I need to bundle the code since I don't  want to publish that folder. I was hoping to avoid using rollup or webpack since I'm just creating a library.

Another option might be to have separate npm packages published that are identical except are compiled to es5.

My (possibly flawed) thinking on the structure was having `main` pointing to commonjs es5 code and `module` pointing to es6 es module based code. I think this would be fine for the browser because bundlers like webpack will use the `module` field by default. Unfortunately node will use the `main` field and so will end up with the non-optimal version of the code.

**Comment by postspectacular** [comment]
I don't think you misunderstood at all and I also can't really see another way than publishing separate packages/versions for all these different outputs:

| | Flavor | Module | Target dir | Entry point | Pkg name suffix |
| --- | --- | --- | --- | --- | --- |
| 1 | ES6 | ES2015 | `.` | `"module": "./index.js"` | |
| 4 | ES6 | UMD |  `./lib` | `"main": "./lib/index.js"` |  |
| 3 | ES5 | ES2015 | `.` | `"module": "./index.js"` | `-es5` |
| 4 | ES5 | UMD |  `./lib` | `"main": "./lib/index.js"` | `-es5` |

We would still need to test/figure out how that impacts import paths for individual files within a package. I.e. if all the UMD files are stored under `/lib`, then won't any consumer also need to have the `lib` path segment in their imports to, e.g. `import { Atom } from "@thi.ng/atom/lib/atom"`?

The other major question is how publishing 2 different versions of each package is going to work (at all) with the current `lerna` workflow / publish script. I really do not want to give up on this, since it makes publishing the entire mono-repo possible. The only way I can think of is to have a publish script for the ES5 versions only and which runs after the `lerna publish` command is finished. This script should:

1) retrieve the tags from the latest commit (which incl. the names of all the packages which have been published)
2) create ES5 versions of the packages in question
3) Create a modified `package.json` (using same version info, but `-es5` as name suffix)
4) Publish each of these ES5 packages individually

Instead of changing the package name via suffix, maybe using `-es5` as version qualifier would be better and less intrusive? E.g. `"@thi.ng/atom": "^1.4.2-es5"` vs `"@thi.ng/atom-es5": "^1.4.2"` - Can you foresee any issues with that?

**Comment by allforabit** [comment]
From a brief look, this looks to be an ideal approach. I really like the idea of adding the es5 as a suffix to the version. This would make switching over to the es6 version very easy as it would just be a simple change to the version in package.json rather than a find and replace to replace all instances of something like `thi.ng/atom-es5`.

WRT the table of output formats, where you have UMD do you not mean commonjs? I think UMD could be another format but would be maybe placed in the dist folder. How about the following?

|   | Flavor | Module | Target dir | Entry point | Pkg name suffix |
|  ------ | ------ | ------ | ------ | ------ | ------ |
|  1 | ES6 | ES2015 | . | module: "./index.js" |  |
|  2 | ES6 | CJS | ./lib | main: "./lib/index.js" |  |
|  3 | ES6 | UMD | ./dist | "unpkg": "dist/<module-name>.js" |  |
|  4 | ES6 | UMD minified | ./dist | - | .min |
|  5 | ES5 | ES2015 | . | module: "./index.js" | -es5 |
|  6 | ES5 | CJS | ./lib | main: "./lib/index.js" | -es5 |
|  7 | ES5 | UMD | ./dist | "unpkg": "dist/<module-name>.js" | -es5 |
|  8 | ES5 | UMD minified | ./dist | - | -es5.min |

**Comment by postspectacular** [comment]
I actually did mean UMD, since I was under the impression it can be consumed by both browser & Node, so as alternative to CommonJS. But I also did mean UMD as TS output module format and *not* as the rollup'd/bundled version. However you're right that there should also be those, both just bundled and bundled/minified versions (3, 4, 7, 8 in your table). Again, I'm not an expert in all those module malarkey, so happy to follow your lead on this...

**Comment by allforabit** [comment]
I can't say I'm an expert on it either! Yes I see what you mean about UMD. Here's an updated table:

|   | Flavor | Module | Target dir | Entry point | Pkg name suffix |
|  ------ | ------ | ------ | ------ | ------ | ------ |
|  1 | ES6 | ES2015 | . | module: "./index.js" |  |
|  2 | ES6 | UMD | ./lib | main: "./lib/index.js" |  |
|  3 | ES6 | Bundled UMD | ./dist | "unpkg": "dist/<module-name>.js" |  |
|  4 | ES6 | Bundled minified UMD | ./dist | - | .min |
|  5 | ES5 | ES2015 | . | module: "./index.js" | -es5 |
|  6 | ES5 | UMD | ./lib | main: "./lib/index.js" | -es5 |
|  7 | ES5 | Bundled UMD | ./dist | "unpkg": "dist/<module-name>.js" | -es5 |
|  8 | ES5 | Bundled minified UMD | ./dist | - | -es5.min |

If that looks good, I'll make a start on putting it together. I may need some help with the publish script but I think I should be able to do all the other stuff.

**Comment by postspectacular** [comment]
Thanks for your help @allforabit! I just pushed what I've got so far to this branch:

https://github.com/thi-ng/umbrella/commit/f5bfd726783c413c2c3fc7e7aabdb16c8a913ea4

Not saying you'll have to base it on this, just for reference :) Good luck!!!!

**Comment by allforabit** [comment]
That looks like a good basis for it, thanks. I will go for something very similar.

**Comment by allforabit** [comment]
I've updated my fork of the repo with a slight variation on what you put together for the builds that includes es5 versions. https://github.com/allforabit/umbrella/blob/feature/rollup/packages/checks/package.json

I've named the build steps as "build:legacy" to indicate that it is only there to support older platforms and to avoid overusing "es". A change that I made and you may have issue with is I've made  the `main` field point to a commonjs build. I'm not sure if it's much use making these UMD modules as they are all very small files and the boilerplate is takes up more than most of the actual code. This reflects advice from here: https://medium.com/@kelin2025/so-you-wanna-use-es6-modules-714f48b3a953 and is also what redux does. 

I've setup the es5 output by making a new `tsconfig.es5.json` file that extends the root `tsconfig.es5.json` as well as `tsconfig.es5.json` file in the root of the package. This is probably littering the directory structure a bit so I can try and find a better way of doing this (maybe just including the config in the command).

**Comment by postspectacular** [comment]
So I finally tried out your version and found a few new issues with the ES5 legacy builds:

1) The bundled UMD result in `dist/checks.js` doesn't really bundle the various files and just includes them like this:

```js
__export(require("./exists-not-null"));
__export(require("./exists"));
__export(require("./has-crypto"));
...
```

Since I have zero experience with UMD (still don't understand **who** is still using UMD in this day and age), I'm not sure if that's supposed to be like that. However, in contrast, the ES6 UMD version **does** include all the functions in this bundled file...

So I did some experimenting and changed the legacy build steps to:

```json
"build:legacy:es": "tsc -p ./tsconfig.es5.json --module es2015 && yarn bundle && tsc -p ./tsconfig.es5.json --declaration",
"build:legacy:commonjs": "tsc -p ./tsconfig.es5.json --declaration --outDir lib --module commonjs",
```

That means we first compile TS to ES2015 module syntax before (and only for) bundling, and then rebuild once more as UMD modules, followed by the CJS build to `/lib` (as before). This causes all the functions to be included properly.

2) Since `@thi.ng/checks` has no dependencies, I've tried applying the same approach to `@thi.ng/atom`, which depends on `@thi.ng/checks` and `@thi.ng/errors` - only to run into new issues related to ES5 and specifying dependencies for the UMD bundle.

E.g. the [error types defined in `@thi.ng/errors` are classes extending `Error`](https://github.com/thi-ng/umbrella/blob/master/packages/errors/src/illegal-arguments.ts). When trying to bundle this package, `rollup` gives me errors relating to the `__extends` polyfill generated by TypeScript:

```
$ (cd packages/errors/ && yarn build:legacy)
yarn run v1.7.0
$ yarn clean && yarn build:legacy:es && yarn build:legacy:commonjs
$ rm -rf *.js *.d.ts .nyc_output build coverage doc
$ tsc -p ./tsconfig.es5.json --module es2015 && yarn bundle && tsc -p ./tsconfig.es5.json --declaration
$ rollup index.js -f umd -o dist/errors.js -n thi.ng.errors && babel-minify -o dist/errors.min.js dist/errors.js

index.js → dist/errors.js...
(!) `this` has been rewritten to `undefined`
https://github.com/rollup/rollup/wiki/Troubleshooting#this-is-undefined
illegal-arguments.js
1: var __extends = (this && this.__extends) || (function () {
                    ^
2:     var extendStatics = Object.setPrototypeOf ||
3:         ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
...and 1 other occurrence
...
```
No idea why or how to fix that... :(

When attempting to bundle `@thi.ng/atom` (or any of the other packages which have dependencies), we will also have to specify their mappings. For that reason I've switched to using a config file like this:

```js
export default {
    input: "index.js",
    context: "window",
    output: {
        file: "dist/atom.js",
        format: "umd",
        name: "thi.ng.atom",
        globals: {
            "@thi.ng/api/mixins/inotify": "thi.ng.api.mixins",
            "@thi.ng/api/mixins/iwatch": "thi.ng.api.mixins",
            "@thi.ng/checks/is-array": "thi.ng.checks",
            "@thi.ng/checks/is-function": "thi.ng.checks",
            "@thi.ng/equiv": "thi.ng.equiv",
            "@thi.ng/errors/illegal-arguments": "thi.ng.errors",
            "@thi.ng/errors/illegal-arity": "thi.ng.errors",
            "@thi.ng/errors/illegal-state": "thi.ng.errors",
            "@thi.ng/paths": "thi.ng.paths",
        }
    }
};
```

*) This is gonna be fun to keep up-to-date in some of the other packages which have more dependencies...
**) It seems the `context` option solves the above "`this` has been rewritten to `undefined`" issue, but can only be specified via config file. the CLI option alternative seems to be ignored...

In summary, I'm seriously doubtful if that effort is all really worth it. That Medium article was way too verbose for me to follow without understanding the reasoning behind this nightmare of hoops to jump through and all the time wondering why, oh why... ES5 support just seems to cause a whole lot of trouble and extra effort for which I personally have zero use for and by now I'm wondering why we do need it? My main use case for these libs are modern browsers and the extensive use of ES6 features throughout all the packages makes me wonder about the performance implications they have in ES5 (with all their polyfills).

Not throwing the towel just yet, but keen to hear your thoughts...

**Comment by pedroteixeira** [comment]
Not sure if it contributes, but in a current typescript project, I ended using a explicit "/src/" in imports for ts source, and leaving `/lib` (or /dist)  `/es` for transpiled ES5 and ES6.

Example:
```
import {bar} from "@repo/package/src/foo";
```

and in the tsconfig.json, I use
```
...
"baseUrl": "./",
"paths": {
    "@repo/package/*": [
          "../package/*"
      ]
}
```

And for some environments (browser and node), also use webpack to generate bundles and can use babel to customize how some dependencies are transpiled or not. 

I didn't see the reason for publishing under different package names - creating different target folders in the same package seems more convenient (for maintainers and consumers). Is it just to minimize package download size?

**Comment by postspectacular** [comment]
@pedroteixeira thanks for this - will try this out, but I still don't understand how we can offer all the options shown in [this table](https://github.com/thi-ng/umbrella/issues/23#issuecomment-399682445) with only one single package, even if we're using dist folders. grateful for any more **concrete** config settings/examples!

I'd really, really like to avoid requiring a bloated/prolonged build setup just get a ES5 version supported out-of-the box, something only very few people will require and which will add complications and extra work and waiting times (due to multiple build targets) for myself & anyone else working on this project...

Having learned more about the IMHO completely ridiculous situation of legacy JS bundling, and considering ES5' days are numbered, I'd personally prefer to just put a note on the READMEs stating that only modern browsers/runtimes with ES6 are supported (like various other projects do too). The libs are heavily using ES6 features and their ES5 versions will be sub-quality due to all the transpiling/polyfilling anyhow...

**Comment by postspectacular** [comment]
So I spent a lot of time again on this over the weekend and think I've found a viable solution. Well, at least it seems to work for me and I've managed to update 8 packages already and created a few dummy examples of user code to test the new package layout:

1. All files are compiled to ES6 modules w/ TS declarations and placed directly in package root (as is currently done, only not output as CommonJS).
2. Use Parcel to build an UMD module (for both Node & browser), placed in `/lib` (plus an additional minified version in same folder)
3. Use Parcel to build a module of the `/test` sources and run `mocha` using this output
4. Add/update `module` & `main` keys in `package.json`
5. Add `browserlist` setting to avoid any form of further transpilation in the UMD modules resulting from step 2. My thinking here is that these can always be transpiled by users themselves if needed, but should just work in all recent versions of the 3 major browsers...

You can follow progress of these changes on this branch:
https://github.com/thi-ng/umbrella/tree/feature/multi-output

So far only these packages have been updated (I started with those with none / few dependencies):

- api
- atom
- bench
- binary
- checks
- equiv
- errors
- paths

**Comment by postspectacular** [comment]
Another update (and recording here only for posterity's sake) and slight adjustment:

I updated all packages, tests & examples and merged everything back into `develop` branch:
https://github.com/thi-ng/umbrella/compare/3df2930...ab15910

Each package now has these build scripts defined in `package.json`, e.g. here for the atom package:

```js
"scripts": {
    "build": "yarn clean && yarn build:es6 && yarn build:bundle",
    "build:es6": "tsc --declaration",
    "build:bundle": "../../scripts/bundle-module atom api checks equiv errors paths",
    "test": "rimraf build && tsc -p test/tsconfig.json && nyc mocha build/test/*.js",
    "clean": "rimraf *.js *.d.ts .nyc_output build coverage doc lib",
    "cover": "yarn test && nyc report --reporter=lcov",
    "doc": "typedoc --mode modules --out doc src",
    "pub": "yarn build && yarn publish --access public"
}
```

The `bundle-module` script itself is here:
https://github.com/thi-ng/umbrella/blob/develop/scripts/bundle-module

It uses `rollup` to produce bundled CJS & UMD versions of the compiled ES6 modules. The latter (along with their respective TS `.d.ts` files) are placed in the package root dir, the bundled outputs are written to `/lib`. Also should mention that the CJS bundle is non-minified, whereas the UMD version is run through `terser`.

Both of these bundled versions **do not** include any dependencies. These are always declared as external...

Each package also has these fields defined:

```js
"module": "./index.js",
"main": "./lib/index.js",
"umd:main": "./lib/index.umd.js",
"typings": "./index.d.ts",
```

Additionally, I've added these settings to avoid `rollup` generating ES5 transpiled bundles and help with tree shaking. **Therefore, all outputs are still using ES6 syntax (regardless of module format)**:

```js
"browserslist": [
    "since 2018-07"
],
"browser": {
    "process": false,
    "setTimeout": false
},
"sideEffects": false
```

As a result, all rebuilt examples are now about 30-50% smaller. 🎉 

I've also tested bundling all ES6 files into a single output (like the other 2 options), however, this results in about 2x larger results when building the examples, so I've stuck w/ separate files for now...

### Breaking changes

As a result of these updates, the new package formats and depending on if you require CommonJS or UMD support, I recommend to **update all `import` declarations to only refer to the package name and not any internal files:**

E.g. Instead of:

```ts
import { map } from "@thi.ng/transducers/xform/map";
```
Only use:

```ts
import { map } from "@thi.ng/transducers";
```

I'm sure most people will welcome that change anyhow and it will also allow for future internal package restructuring without breaking user code... everything else should work just as before (related [commit](https://github.com/thi-ng/umbrella/commit/b54b703d4d694f86845925de40d919f7451b5e2b))

Re: examples - I've added a new wiki page with updated build instructions and some potential new pitfalls with a few examples (related to a Parcel issue):

https://github.com/thi-ng/umbrella/wiki/Example-build-instructions

I'm planning to merge & release these changes over the coming weekend. So if anyone has objections, please do let me know asap...

**Comment by postspectacular** [comment]
changes have now been published as major new versions of all packages. Any feedback is very welcome. I especially would love to hear from UMD users if those bundles are working fine (they do seem to in my testing).

You can include UMD versions like this:
https://unpkg.com/@thi.ng/geom@1.0.0/lib/index.umd.js

But note, that all umbrella UMD packages do **not** include any dependencies...

Closing this for now!

**Comment by gavinpc-mindgrub** [comment]
Many thanks.  I am using the UMD bundles with RequireJS and they work fine.

I first started using thi.ng early this year and the lack of an AMD-compatible distribution format was the only notable pain point.  I watched the repo and this the first notification I received.  It is all truly amazing work and a great gift to the community.

**Comment by postspectacular** [comment]
Thanks for letting me know @gavinpc-mindgrub - and thanks again for kind words! (plus a reminder to email you 👍 ) Since I will soon be embarking on an Amazon Lambda journey, I might have to look into ways to also provide ES5 versions too (since Lambda only seems to support Node 8), but hoping I can get away w/ bundling and only needing to transpile those artefacts.
